master/internal/api_experiment.go

```diff
@@ -2337,3 +2338,109 @@ func (a *apiServer) CreateTrial(
 
 	return resp, nil
 }
+
+func (a *apiServer) PutExperimentLabel(ctx context.Context,
+	req *apiv1.PutExperimentLabelRequest,
+) (*apiv1.PutExperimentLabelResponse, error) {
+	curUser, _, err := grpcutil.GetUser(ctx)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to get the user: %s", err)
+	}
+
+	exp, err := a.getExperiment(ctx, *curUser, int(req.ExperimentId))
+	if err != nil {
+		return nil, err
+	}
+
+	modelExp, err := model.ExperimentFromProto(exp)
+	if err != nil {
+		return nil, err
+	}
+
+	if err = exputil.AuthZProvider.Get().CanEditExperimentsMetadata(
+		ctx, *curUser, modelExp); err != nil {
+		return nil, status.Errorf(codes.PermissionDenied, err.Error())
+	}
+
+	// Labels should be unique.
+	expLabels := set.New[string]()
+	for _, label := range exp.Labels {
+		expLabels.Insert(label)
+	}
+	if expLabels.Contains(req.Label) {
+		return &apiv1.PutExperimentLabelResponse{Labels: exp.Labels}, nil
+	}
+	expLabels.Insert(req.Label)
+
+	exp.Labels = expLabels.ToSlice()
+
+	type experimentPatch struct {
+		Labels []string `json:"labels"`
+	}
+	patch := experimentPatch{
+		Labels: exp.Labels,
+	}
+	marshalledPatches, patchErr := json.Marshal(patch)
+	if patchErr != nil {
+		return nil, status.Errorf(codes.Internal, "failed to marshal experiment patch")
+	}
+
+	_, err = a.m.db.RawQuery(
+		"patch_experiment", exp.Id, marshalledPatches, exp.Notes)
+	if err != nil {
```

Comment:
running into the same problem here as the original problem, but with the span of time it can happen in from line 2350 to 2390. I would just write a new query to patch the labels, or do all this in a transaction.

Reply:
At least before, it is semi obvious that determined would not be doing transactions over multiple API calls in the flow GET /experiments/:id, add my label, PATCH /experiments/:id. Losing updates in the course of a single API call is a less obvious/acceptable.

Reply:
you're totally right, i tried to follow `PatchExperiments` and blindly assumed that the transaction was somehow magically handled 🤦 will fix this.

Reply:
fixed